Cython/Compiler/ExprNodes.py

```diff
@@ -2008,11 +2008,20 @@ def declare_from_annotation(self, env, as_target=False):
                           "'%s' cannot be specialized since its type is not a fused argument to this function" %
                           self.name)
                     atype = error_type
-
-            entry = self.entry = env.declare_var(name, atype, self.pos, is_cdef=not as_target)
+            kwds = {}
+            if 'dataclasses.dataclass' in env.directives:
+                # handle "frozen" directive - full inspection of the dataclass directives happens
+                # in Dataclass.py
+                frozen_directive = env.directives['dataclasses.dataclass']
+                if frozen_directive:
+                    frozen_directive = frozen_directive[1].get('frozen', None)
+                is_frozen = frozen_directive and frozen_directive.is_literal and frozen_directive.value
+                kwds = {'visibility': 'readonly?' if is_frozen else 'public?'}
```

Comment:
I'd rather have this as an explicit marker that makes the property generation optional or sets it as internally requested.

OTOH, we have a type here already, don't we? Why not look at it? Even if the decision is non-trivial, it can be taken by the type itself.

Reply:
Not exactly what you suggested, but I think it gets simplified a lot by making the decision here in ExprNodes (since we have all the information needed).